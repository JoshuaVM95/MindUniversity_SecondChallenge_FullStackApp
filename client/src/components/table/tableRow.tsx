import React, { useState } from "react";
import {
	Box,
	Collapse,
	IconButton,
	Table,
	TableBody,
	TableCell,
	TableHead,
	TableRow,
	Typography
} from "@material-ui/core";
import { KeyboardArrowDown, KeyboardArrowUp } from "@material-ui/icons";

type stringObject = { [key: string]: string };

export interface RowObject {
	[key: string]: string | unknown;
	collapsableTableData?: stringObject[];
}

interface RowProps<T extends RowObject> {
	row: T;
	collapsableTableTitle?: string;
	collapsableTableHeaders?: string[];
	onCollapse?(show: boolean): void;
	onRowSelected(isRowSelected: boolean): void;
}

export const Row = <T extends RowObject>({
	row,
	collapsableTableTitle,
	collapsableTableHeaders,
	onCollapse,
	onRowSelected
}: RowProps<T>): React.ReactElement => {
	const [showCollapsableData, setShowCollapsableData] = useState<boolean>(false);
	const [isRowSelected, setIsRowSelected] = useState<boolean>(false);

	return (
		<>
			<TableRow
				sx={{ "& > *": { borderBottom: "unset" } }}
				hover
				onClick={() => {
					onRowSelected(!isRowSelected);
					setIsRowSelected(!isRowSelected);
				}}
				role="checkbox"
				aria-checked={isRowSelected}
				tabIndex={-1}
				selected={isRowSelected}
			>
				<TableCell>
					<IconButton
						aria-label="expand row"
						size="small"
						onClick={(event) => {
							event.stopPropagation();
							if (onCollapse) onCollapse(!showCollapsableData);
							setShowCollapsableData(!showCollapsableData);
						}}
					>
						{showCollapsableData ? <KeyboardArrowUp /> : <KeyboardArrowDown />}
					</IconButton>
				</TableCell>
				{Object.keys(row)
					.filter((rowKey) => typeof row[rowKey] !== "object")
					.map((rowKey) => {
						return (
							<TableCell key={rowKey} align="center">
								{row[rowKey] as string}
							</TableCell>
						);
					})}
			</TableRow>
			{row.collapsableTableData !== undefined && collapsableTableHeaders && (
				<TableRow>
					<TableCell style={{ paddingBottom: 0, paddingTop: 0 }} colSpan={6}>
						<Collapse in={showCollapsableData} timeout="auto" unmountOnExit>
							<Box sx={{ margin: 1 }}>
								<Typography variant="h6" gutterBottom component="div">
									{collapsableTableTitle}
								</Typography>
								<Table size="small" aria-label="purchases">
									<TableHead>
										<TableRow>
											{collapsableTableHeaders.map((header) => (
												<TableCell key={header} align="center">
													{header}
												</TableCell>
											))}
										</TableRow>
									</TableHead>
									<TableBody>
										{row.collapsableTableData &&
											row.collapsableTableData.map((detail, index) => {
												return (
													<TableRow key={index}>
														{Object.keys(detail).map((detailKey) => {
															const detailRow = detail[detailKey];
															return (
																<TableCell key={detailKey} align="center">
																	{detailRow}
																</TableCell>
															);
														})}
													</TableRow>
												);
											})}
									</TableBody>
								</Table>
							</Box>
						</Collapse>
					</TableCell>
				</TableRow>
			)}
		</>
	);
};
