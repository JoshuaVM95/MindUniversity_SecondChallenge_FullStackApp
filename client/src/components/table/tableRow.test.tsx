import React from "react";
import { Row } from "./tableRow";
import { render, fireEvent, getByTestId } from "@testing-library/react";

describe("TableRow tests", () => {
	const collapse = jest.fn();
	const selected = jest.fn();
	const edit = jest.fn();
	const renderRow = () => {
		const row = {
			first: "1"
		};
		return render(
			<Row
				row={row}
				collapsableTableTitle="Title"
				collapsableTableHeaders={["My secondary header"]}
				onCollapse={collapse}
				onRowSelected={selected}
				onEdit={edit}
			/>
		);
	};
	it("renders correctly", () => {
		expect(renderRow().asFragment()).toMatchSnapshot();
	});
	it("renders the text prop correctly", () => {
		const row = renderRow().asFragment();
		expect(row).toHaveTextContent("1");
	});
	it("triggers click", () => {
		const row = renderRow().container;
		fireEvent.click(getByTestId(row, "KeyboardArrowDownIcon"));
		expect(collapse).toHaveBeenCalled();
	});
});
