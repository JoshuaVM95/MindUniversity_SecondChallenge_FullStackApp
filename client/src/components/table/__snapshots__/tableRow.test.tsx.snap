// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableRow tests renders correctly 1`] = `
<DocumentFragment>
  <table
    class="MuiTable-root css-rqglhn-MuiTable-root"
  >
    <tbody
      class="MuiTableBody-root css-apqrd9-MuiTableBody-root"
    >
      <tr
        aria-checked="false"
        class="MuiTableRow-root MuiTableRow-hover css-hw5v8q-MuiTableRow-root"
        role="checkbox"
        tabindex="-1"
      >
        <td
          class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
        >
          <button
            aria-label="expand row"
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall css-1pe4mpk-MuiButtonBase-root-MuiIconButton-root"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="KeyboardArrowDownIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M7.41 8.59 12 13.17l4.59-4.58L18 10l-6 6-6-6 1.41-1.41z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </td>
        <td
          class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1yhpg23-MuiTableCell-root"
        >
          1
        </td>
      </tr>
      <tr
        class="MuiTableRow-root css-1q1u3t4-MuiTableRow-root"
      >
        <td
          class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
          colspan="6"
          style="padding-bottom: 0px; padding-top: 0px;"
        />
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;
